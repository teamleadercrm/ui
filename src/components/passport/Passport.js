import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import theme from './theme.css';
import Avatar from '../avatar';
import Box from '../box';
import Icon from '../icon';
import Link from '../link';
import Popover from '../popover';
import { Heading3, TextBody } from '../typography';

/** @type {React.ComponentType<any>} */
class Passport extends PureComponent {
  renderDescription = () => {
    const { description } = this.props;

    if (!description) {
      return null;
    }

    if (Array.isArray(description)) {
      return (
        <>
          {description.map((child, index) => (
            <TextBody color="teal" key={index}>
              {child}
            </TextBody>
          ))}
        </>
      );
    } else {
      return <TextBody color="teal">{description}</TextBody>;
    }
  };

  renderTitle = () => {
    const { title } = this.props;

    return (
      <Heading3 className={cx(theme['prevent-overflow'], theme['prevent-wrap'], theme['show-ellipsis'])} color="teal">
        {typeof title === 'string' ? title : <Link {...title} inherit={false} />}
      </Heading3>
    );
  };

  render() {
    const { avatar, className, lineItems, ...others } = this.props;

    return (
      <Popover {...others} backdrop="transparent" className={cx(theme['passport'], className)}>
        <Box padding={3}>
          <Box display="flex">
            <Box flex="48px 0 0" paddingRight={3}>
              <Avatar {...avatar} size="small" />
            </Box>
            <Box className={theme['prevent-overflow']}>
              {this.renderTitle()}
              {this.renderDescription()}
            </Box>
          </Box>
          {lineItems && (
            <Box marginTop={3}>
              {lineItems.map(({ icon, ...lineItem }, index) => (
                <Box alignItems="flex-start" display="flex" key={index}>
                  <Box display="flex" flex="48px 0 0" justifyContent="center" paddingRight={3}>
                    {icon && (
                      <Icon
                        color={lineItem.href || lineItem.onClick ? 'aqua' : 'teal'}
                        tint={lineItem.href || lineItem.onClick ? 'dark' : 'darkest'}
                        marginTop={1}
                      >
                        {icon}
                      </Icon>
                    )}
                  </Box>
                  <TextBody
                    className={cx(theme['prevent-overflow'], theme['prevent-wrap'], theme['show-ellipsis'])}
                    color="teal"
                  >
                    {lineItem.href || lineItem.onClick ? <Link {...lineItem} inherit={false} /> : lineItem.children}
                  </TextBody>
                </Box>
              ))}
            </Box>
          )}
        </Box>
      </Popover>
    );
  }
}

Passport.propTypes = {
  /** Object containing the props of an Avatar. */
  avatar: PropTypes.object.isRequired,
  /** The class names for the wrapper to apply custom styling. */
  className: PropTypes.string,
  /** The description of the Passport. */
  description: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
  /** Array of objects containing the props of a Link. */
  lineItems: PropTypes.array,
  /** The title of the Passport (can be a string of an object containing the props of a Link). */
  title: PropTypes.oneOfType([PropTypes.object, PropTypes.string]).isRequired,
};

export default Passport;
