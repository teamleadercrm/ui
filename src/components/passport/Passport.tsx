import React, { ReactNode } from 'react';
import cx from 'classnames';
import theme from './theme.css';
import Avatar from '../avatar';
import Box from '../box';
import Icon from '../icon';
import Link from '../link';
import Popover from '../popover';
import { Heading3, TextBody } from '../typography';
import { AvatarProps } from '../avatar/Avatar';
import { LinkProps } from '../link/Link';
import { GenericComponent } from '../../@types/types';
import { BoxProps } from '../box/Box';

interface LinkPropsAndIcon extends LinkProps {
  icon?: ReactNode;
}
interface PassportProps extends Omit<BoxProps, 'className'> {
  /** Object containing the props of an Avatar. */
  avatar: AvatarProps;
  /** The class names for the wrapper to apply custom styling. */
  className?: string;
  /** The description of the Passport. */
  description?: string | string[];
  /** Array of objects containing the props of a Link. */
  lineItems?: LinkPropsAndIcon[];
  /** The title of the Passport (can be a string of an object containing the props of a Link). */
  title: string | LinkPropsAndIcon;
}

const Passport: GenericComponent<PassportProps> = ({ avatar, className, description, lineItems, title, ...others }) => {
  const renderDescription = () => {
    if (!description) {
      return null;
    }

    if (Array.isArray(description)) {
      return (
        <>
          {description.map((child, index) => (
            <TextBody color="teal" key={index}>
              {child}
            </TextBody>
          ))}
        </>
      );
    } else {
      return <TextBody color="teal">{description}</TextBody>;
    }
  };

  const renderTitle = () => {
    return (
      <Heading3 className={cx(theme['prevent-overflow'], theme['prevent-wrap'], theme['show-ellipsis'])} color="teal">
        {typeof title === 'string' ? title : <Link {...title} inherit={false} />}
      </Heading3>
    );
  };

  return (
    <Popover {...others} backdrop="transparent" className={cx(theme['passport'], className)}>
      <Box padding={3}>
        <Box display="flex">
          <Box flex="48px 0 0" paddingRight={3}>
            <Avatar {...avatar} size="small" />
          </Box>
          <Box className={theme['prevent-overflow']}>
            {renderTitle()}
            {renderDescription()}
          </Box>
        </Box>
        {lineItems && (
          <Box marginTop={3}>
            {lineItems.map(({ icon, ...lineItem }, index) => (
              <Box alignItems="flex-start" display="flex" key={index}>
                <Box display="flex" flex="48px 0 0" justifyContent="center" paddingRight={3}>
                  {icon && (
                    <Icon
                      color={lineItem.href || lineItem.onClick ? 'aqua' : 'teal'}
                      tint={lineItem.href || lineItem.onClick ? 'dark' : 'darkest'}
                      marginTop={1}
                    >
                      {icon}
                    </Icon>
                  )}
                </Box>
                <TextBody
                  className={cx(theme['prevent-overflow'], theme['prevent-wrap'], theme['show-ellipsis'])}
                  color="teal"
                >
                  {lineItem.href || lineItem.onClick ? <Link {...lineItem} inherit={false} /> : lineItem.children}
                </TextBody>
              </Box>
            ))}
          </Box>
        )}
      </Box>
    </Popover>
  );
};

export default Passport;
