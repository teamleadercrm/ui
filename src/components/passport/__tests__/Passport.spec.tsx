import React from 'react';
import { render } from '@testing-library/react';
import Passport from '../Passport';

describe('Component - Passport', () => {
  it('renders', async () => {
    const screen = render(
      <Passport
        title="Dwight Schrute"
        description="Assistant to the regional manager"
        avatar={{ imageUrl: 'https://foobar.com/dwight.png', fullName: 'Dwight Schrute' }}
      />,
    );
    expect(screen.baseElement).toMatchSnapshot();
  });
});
