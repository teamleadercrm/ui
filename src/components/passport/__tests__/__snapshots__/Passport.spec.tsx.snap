// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Passport renders 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    class="wrapper neutral lightest is-entering"
    style="z-index: 300;"
  >
    <div
      class="overlay transparent overlay"
      data-teamleader-ui="overlay"
    >
      <div>
        <div>
          <div
            tabindex="0"
          />
          <div
            class="box popover passport"
            data-teamleader-ui="popover"
            style="left: 0px; top: 0px; max-width: 50vw; min-width: 180px;"
          >
            <div
              class="box display-flex flex-direction-column inner"
              style="flex: 1 1 auto;"
            >
              <div
                class="box padding-bottom-3 padding-left-3 padding-right-3 padding-top-3"
                data-teamleader-ui="passport"
              >
                <div
                  class="box display-flex"
                >
                  <div
                    class="box padding-right-3"
                    style="flex: 0 0px;"
                  >
                    <div
                      class="box wrapper is-small is-circle"
                    >
                      <div
                        class="avatar avatar-image"
                        data-teamleader-ui="avatar-image"
                      >
                        <img
                          alt="Dwight Schrute"
                          class="image"
                          src="https://foobar.com/dwight.png"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="box prevent-overflow"
                  >
                    <h3
                      class="box heading heading-3 darkest teal prevent-overflow prevent-wrap show-ellipsis"
                      data-teamleader-ui="heading"
                    >
                      Dwight Schrute
                    </h3>
                    <p
                      class="box text text-body darkest teal"
                      data-teamleader-ui="text"
                    >
                      Assistant to the regional manager
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
